akamai/cache-client.go

```diff
@@ -0,0 +1,255 @@
+	)
+	bodyHash := sha256.Sum256(body)
+	tbs := fmt.Sprintf(
+		"%s\t%s\t%s\t%s\t%s\t%s\t%s",
+		"POST",
+		cpc.apiScheme,
+		cpc.apiHost,
+		apiPath,
+		"", // We don't need to send any signed headers so this can be blank
+		base64.StdEncoding.EncodeToString(bodyHash[:]),
+		header,
+	)
+
+	h := hmac.New(sha256.New, []byte(cpc.clientSecret))
+	h.Write([]byte(timestamp))
+	key := base64.StdEncoding.EncodeToString(h.Sum(nil))
+
+	h = hmac.New(sha256.New, []byte(key))
+	h.Write([]byte(tbs))
+	sig := base64.StdEncoding.EncodeToString(h.Sum(nil))
+
+	return fmt.Sprintf("%ssignature=%s", header, sig), nil
+}
+
+func (cpc *CachePurgeClient) generateURLs(der []byte) ([]string, error) {
+	cert, err := x509.ParseCertificate(der)
+	if err != nil {
+		return nil, err
+	}
+
+	req, err := ocsp.CreateRequest(cert, cpc.issuer, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	urls := []string{}
+	for _, ocspServer := range cert.OCSPServer {
+		if !strings.HasSuffix(ocspServer, "/") {
+			ocspServer += "/"
+		}
+		urls = append(
+			urls,
+			fmt.Sprintf("%s%s", ocspServer, url.QueryEscape(base64.StdEncoding.EncodeToString(req))),
+		)
+	}
+	return urls, nil
+}
+
+func (cpc *CachePurgeClient) purge(der []byte) error {
+	urls, err := cpc.generateURLs(der)
+	if err != nil {
+		return errFatal(err.Error())
+	}
+	purgeReq := purgeRequest{urls}
+	reqJSON, err := json.Marshal(purgeReq)
+	if err != nil {
+		return errFatal(err.Error())
+	}
+
+	req, err := http.NewRequest(
+		"POST",
+		fmt.Sprintf("%s%s", cpc.apiEndpoint, purgePath),
+		bytes.NewBuffer(reqJSON),
+	)
+	if err != nil {
+		return errFatal(err.Error())
+	}
+
+	nonce := make([]byte, 16)
+	_, err = rand.Read(nonce)
+	if err != nil {
+		return err
+	}
+	authHeader, err := cpc.constructAuthHeader(
+		req,
+		reqJSON,
+		purgePath,
+		base64.StdEncoding.EncodeToString(nonce),
+	)
+	if err != nil {
+		return errFatal(err.Error())
+	}
+	req.Header.Set("Authorization", authHeader)
+	req.Header.Set("Content-Type", "application/json")
+
+	rS := cpc.clk.Now()
+	resp, err := cpc.client.Do(req)
+	cpc.stats.TimingDuration("CCU.PurgeRequestLatency", time.Since(rS), 1.0)
+	if err != nil {
+		return err
+	}
+	if resp.Body == nil {
+		return fmt.Errorf("No response body")
+	}
+	defer resp.Body.Close()
+	body, err := ioutil.ReadAll(resp.Body)
+	if err != nil {
+		return err
+	}
+
+	var purgeInfo purgeResponse
+	err = json.Unmarshal(body, &purgeInfo)
+	if err != nil {
+		return err
+	}
+	if purgeInfo.HTTPStatus != 201 || resp.StatusCode != 201 {
+		return fmt.Errorf("Incorrect HTTP status code")
+	}
+
+	cpc.log.Info(fmt.Sprintf(
+		"Sent successful purge request purgeID: %s, purge expected in: %ds, for URLs: %s",
+		purgeInfo.PurgeID,
+		purgeInfo.EstimatedSeconds,
+		urls,
+	))
+
+	return nil
+}
+
+// Purge sends a purge request to the configured API endpoint
+func (cpc *CachePurgeClient) Purge(der []byte) error {
+	successful := false
+	for i := 0; i <= cpc.retries; i++ {
+		if i > 0 {
+			time.Sleep(cpc.retryBackoff)
```

Comment:
consider and implementation like for backoff https://github.com/grpc/grpc-go/blob/master/rpc_util.go#L311